src/server/pkg/worker/branch_set.go

```diff
@@ -0,0 +1,245 @@
+							rootCommits[commit.Repo.Name] = make(map[string]bool)
+						}
+						rootCommits[commit.Repo.Name][commit.ID] = true
+					}
+					for _, commitInfo := range set {
+						setRootCommit(commitInfo.Commit)
+						for _, provCommit := range commitInfo.Provenance {
+							setRootCommit(provCommit)
+						}
+					}
+					// ok tells us if it's ok to spawn a job with this
+					// commit set.
+					for _, rootInput := range rootInputs {
+						if len(rootCommits[rootInput.Repo]) != 1 {
+							return false
+						}
+					}
+					return true
+				}); set != nil {
+					bs := &branchSet{
+						NewBranch: i,
+					}
+					for _, commitInfo := range set {
+						bs.Branches = append(bs.Branches, &pfs.Branch{
+							Name: input.Branch,
+							Head: commitInfo.Commit,
+						})
+					}
+					select {
+					case <-ctx.Done():
+						commitSetsMutex.Unlock()
+						return
+					case ch <- bs:
+					}
+				}
+				commitSetsMutex.Unlock()
+			}
+		}()
+	}
+
+	f := &branchSetFactoryImpl{
+		cancel: cancel,
+		ch:     ch,
+	}
+
+	return f, nil
+}
+
+// considerEachSet runs a function on every commit set, starting with the
+// most recent one.  If the function returns true, then considerEachSet
+// removes the commit sets that are older than the current one and returns
+// the current one.
+// i is the index of the input from which we just got a new commit.  Since
+// it's the "triggering" input, we know that we only have to consider commit
+// sets that include the triggering commit since other commit sets must have
+// already been considered in previous runs.
+func considerEachSet(commitSets [][]*pfs.CommitInfo, i int, f func(commitSet []*pfs.CommitInfo) bool) []*pfs.CommitInfo {
+	numCommitSets := 1
+	for j, commits := range commitSets {
+		if i != j {
+			numCommitSets *= len(commits)
+		}
+	}
+	for j := numCommitSets - 1; j >= 0; j-- {
+		numCommitSets := j
+		var commitSet []*pfs.CommitInfo
+		var indexes []int
+		for k, commits := range commitSets {
+			var index int
+			if k == i {
+				index = len(commits) - 1
+			} else {
+				index = numCommitSets % len(commits)
+				numCommitSets /= len(commits)
+			}
+			indexes = append(indexes, index)
+			commitSet = append(commitSet, commits[index])
+		}
+		if f(commitSet) {
+			// Remove older commit sets
+			for k, index := range indexes {
+				commitSets[k] = commitSets[k][index:]
+			}
+			return commitSet
+		}
+	}
+	return nil
+}
+
+// directInputs returns the inputs that should trigger the pipeline.  Inputs
+// from the same repo/branch are de-duplicated.
+func (a *APIServer) directInputs(ctx context.Context) []*pps.AtomInput {
+	repoSet := make(map[string]bool)
+	var atomInputs []*pps.AtomInput
+	pps.VisitInput(a.pipelineInfo.Input, func(input *pps.Input) {
+		if input.Atom != nil {
+			if repoSet[input.Atom.Repo] {
+				return
+			}
+			repoSet[input.Atom.Repo] = true
+			atomInputs = append(atomInputs, input.Atom)
+		}
+	})
+	return atomInputs
+}
```

Comment:
Purely in the interest of not having too many words for the same concept, how would you feel about calling this `trueInputs` that's what we'd been using before.